### ISSUE TITLE
Use it in flutter

### ISSUE THREAD

**Issue by xAffan** [Foundation-Devices/tor#26]
How do you actually use the socks5 proxy in flutter? It seems impossible!

**Comment by icota** [comment]
Hi @xAffan. We use this plugin in https://github.com/Foundation-Devices/envoy and a Rust dependency of ours (https://github.com/bitcoindevkit) looks up onion addresses without fail.

We have never actually tried accessing .onions from Dart code. Maybe @sneurlax did?

**Comment by xAffan** [comment]
I haven't found a way to lookup onion addresses from dart. Both of you seem much more experienced than me, therefore, I would be very happy if you guided me how to lookup onion addresses preferably from dart, if not, then from bitcoindevkit dependency(?)

**Comment by sneurlax** [comment]
I do get onion responses when running arti locally on desktop, but haven't thoroughly tested it when arti's running as bundled in this package.  We are still waiting on arti to release full, secure/private onion addresses and will probably finish integration after that.

@xAffan I had to write a custom SOCKS5 package for communication with electrumx because nothing else I found worked/supported sockets properly.  You might have success finding something else that works, but this may work or we may just have to patch it for SOCKS5H: https://github.com/cypherstack/socks_socket

**Comment by xAffan** [comment]
It does actually support .onion. However, I am unable to connect to it properly using flutter, I just need to be able to use socks5h with some http client. I succeeded with using socks5 because someone had made a package implementating socks5 but I suspect it is not socks5h and does DNS locally.

**Comment by sneurlax** [comment]
May I ask what you're trying to connect to?  And/or what SOCKS5 package you're using?

I've tried these SOCKS5 dart packages which work for basic functionality but didn't totally work for what we needed (with ElectrumX use with sockets)--unrelated to your .onion address issues, so I'd check them out in case they might work:

- [LacticWhale/socks_dart](https://github.com/LacticWhale/socks_dart)
- [v0l/socks5](https://github.com/v0l/socks5)
- [alexd1971's socks5 example](https://gist.github.com/alexd1971/2ac145546ee72f713e0c9fb6acddf425)

This is what I wrote when none of the above worked perfectly for what we needed, so maybe this will help?

- [cypherstack/socks_socket](https://github.com/cypherstack/socks_socket)
If none of the above work, if you can provide a working commandline example of a SOCKS5H command/response or an example of a SOCKS5 call that works and a SOCKS5H one that doesn't, that would help me get started on extending that socks_socket package to support SOCKS5H.

**Comment by sneurlax** [comment]
PS if none of the upper packages work, for using socks_socket, the links from [here](https://github.com/cypherstack/socks_socket/issues/5) might help.  I need to finish socks_socket :D

**Comment by xAffan** [comment]
I was using https://github.com/LacticWhale/socks_dart
It worked but couldn't connect to onion sites due to DNS lookup failure I believe.
I will now test with your package to see if it works and the other two packages as well.

**Comment by nns52k** [comment]
[Tor Project: FAQ](https://2019.www.torproject.org/docs/faq.html.en#AccessOnionServices) basically says we should use SOCKS 4a. If we use SOCKS 5, it won't work. It is apparently on purpose to avoid DNS lookups.

I tried it a few years ago in C code, doing SOCKS 4a handshake myself. It works. I also use Boost.Asio and libcurl respectively in my C++ codes. They all work as long as we stick to SOCKS 4a.

I'm new to Flutter, and haven't tried using SOCKS 4a in Dart code to connect onion service yet. If using SOCKS 4a works in Dart/Flutter code, please let us know. Thanks.

**Comment by xAffan** [comment]
> [Tor Project: FAQ](https://2019.www.torproject.org/docs/faq.html.en#AccessOnionServices) basically says we should use SOCKS 4a. If we use SOCKS 5, it won't work. It is apparently on purpose to avoid DNS lookups.
> 
> I tried it a few years ago in C code, doing SOCKS 4a handshake myself. It works. I also use Boost.Asio and libcurl respectively in my C++ codes. They all work as long as we stick to SOCKS 4a.
> 
> I'm new to Flutter, and haven't tried using SOCKS 4a in Dart code to connect onion service yet. If using SOCKS 4a works in Dart/Flutter code, please let us know. Thanks.

It is possible to use SOCKS 4a.
https://github.com/utopicnarwhal/flibusta-mobile/blob/master/lib/services/http_client/dio_http_client_adapters/socks_http_client_adapter.dart
I would have gone ahead and used it but it is dart 2.x code. Anyways, this project did use SOCKS 4a.

**Comment by sneurlax** [comment]
Hey @xAffan, did you ever try socks_socket.dart?

I had to use it again today for connecting electrum_adapter via Tor.  If you can report some issues/errors, I'll try and fix them.  Or if you provide some minimal reproducible code showing what you want to succeed that doesn't, I'll try and extend socks_socket to accomplish that.

You can use it like
```
import 'package:tor/socks_socket.dart';

<some async func>:
await Tor.init();
await Tor.instance.start();

SOCKSSocket socksSocket = await SOCKSSocket.create(
  proxyHost: InternetAddress.loopbackIPv4,
  proxyPort: Tor.instance.port,
  sslEnabled: true,
);
await socksSocket?.connect();
await _socksSocket?.connectTo(<destHost>, <destPort>);
```

**Comment by xAffan** [comment]
> Hey @xAffan, did you ever try socks_socket.dart?
> 
> I had to use it again today for connecting electrum_adapter via Tor. If you can report some issues/errors, I'll try and fix them. Or if you provide some minimal reproducible code showing what you want to succeed that doesn't, I'll try and extend socks_socket to accomplish that.
> 
> You can use it like
> 
> ```
> import 'package:tor/socks_socket.dart';
> 
> <some async func>:
> await Tor.init();
> await Tor.instance.start();
> 
> SOCKSSocket socksSocket = await SOCKSSocket.create(
>   proxyHost: InternetAddress.loopbackIPv4,
>   proxyPort: Tor.instance.port,
>   sslEnabled: true,
> );
> await socksSocket?.connect();
> await _socksSocket?.connectTo(<destHost>, <destPort>);
> ```

I would use this but I am not quite sure how to use HTTP protocol with this? I need to make some GET and POST requests, and I am not exactly sure how to.

**Comment by sneurlax** [comment]
@xAffan if you don't need it for electrum servers, maybe my socks_socket is not for you.  But see https://github.com/Foundation-Devices/tor/blob/main/lib/socks_socket.dart#L321 (the `sendServerFeaturesCommand` method) for a close example: instead of sending a JSON command, you'll send a raw GET or POST request.  For example:
```
Future<void> makeGetRequest(SOCKSSocket socksSocket) async {
  // Domain to make a GET request to
  var domain = 'example.com';

  // GET request path
  var path = '/api/data';

  // Prepare the GET request
  var getRequest = 'GET $path HTTP/1.1\r\n'
      'Host: $domain\r\n'
      'Connection: close\r\n\r\n';

  // Send the GET request
  socksSocket.write(getRequest);

  // Read and print the response
  await socksSocket.inputStream
      .transform(utf8.decoder)
      .transform(LineSplitter())
      .forEach(print);
}

Future<void> makePostRequest(SOCKSSocket socksSocket) async {
  // Domain to make a POST request to
  var domain = 'example.com';

  // POST request path
  var path = '/api/data';

  // POST request body
  var requestBody = jsonEncode({
    'key': 'value',
  });

  // Prepare the POST request
  var postRequest = 'POST $path HTTP/1.1\r\n'
      'Host: $domain\r\n'
      'Content-Type: application/json\r\n'
      'Content-Length: ${requestBody.length}\r\n'
      'Connection: close\r\n\r\n'
      '$requestBody';

  // Send the POST request
  socksSocket.write(postRequest);

  // Read and print the response
  await socksSocket.inputStream
      .transform(utf8.decoder)
      .transform(LineSplitter())
      .forEach(print);
}
```
These may not work if you copy and paste them but show the general approach you might take towards making GET and POST requests via socks_socket.

It may help if you can post a small example of exactly what you're trying to do (or something similar enough)

**Comment by xAffan** [comment]
Unfortunately, raw requests won't do.
here's me using ur sample.

Launching lib\main.dart on Windows in debug mode...
√  Built build\windows\x64\runner\Debug\main.exe.
Connecting to VM Service at ws://127.0.0.1:50397/ZLWv9PAmtXU=/ws
flutter: NOW: 2024-04-24 04:15:19.796316
flutter: Instance of Tor created!
Starting proxy!
flutter: Done awaiting; tor should be running
flutter: Starting tor took 261 seconds. Proxy running on port 43787
flutter: 192.42.116.199

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: SocketException: Failed host lookup: 'duckduckgogg42xjoc72x3sjasowoarfbgcmvfimaftt6twagswzczad.onion' (OS Error: No such host is known.

, errno = 11001)
#0      _NativeSocket.lookup.<anonymous closure> (dart:io-patch/socket_patch.dart:520:9)
<asynchronous suspension>
#1      SocksTCPClient.connect (package:socks5_proxy/src/client/socks_tcp_client.dart:66:18)
<asynchronous suspension>
#2      SocksTCPClient.assignToHttpClientWithSecureOptions.<anonymous closure> (package:socks5_proxy/src/client/socks_tcp_client.dart:46:40)
<asynchronous suspension>
#3      _ConnectionTarget.connect.<anonymous closure> (dart:_http/http_impl.dart:2490:32)
<asynchronous suspension>
#4      _HttpClient._openUrl.<anonymous closure> (dart:_http/http_impl.dart:2787:15)
<asynchronous suspension>
#5      _MyAppState.build.<anonymous closure> (package:xclient/main.dart:158:45)
<asynchronous suspension>

**Comment by nns52k** [comment]
@sneurlax It'd be great if we can connect with onion servers, e.g. DuckDuckGo `https://duckduckgogg42xjoc72x3sjasowoarfbgcmvfimaftt6twagswzczad.onion`.

For now, we can only connect to ordinary websites. If we try to connect to an onion server, it will fail. In my case of using [LacticWhale/socks_dart](https://github.com/LacticWhale/socks_dart) to connect an onion router, it throws:
```
SocketException: Failed host lookup: 'duckduckgogg42xjoc72x3sjasowoarfbgcmvfimaftt6twagswzczad.onion' (OS Error: No address associated with hostname, errno = 7)
```

By the way, does your [sneurlax/socks_socket](https://github.com/sneurlax/socks_socket) also provide a wayt to create or adapt a `HttpClient` or `Client`? Classes `HttpClient` and `Client` are from packages `dart:io` and `http` respectively. It'd be more user-friendly if it does. Some Flutter packages, e.g. Supabase, use `Client`.

**Comment by xAffan** [comment]
> @sneurlax It'd be great if we can connect with onion servers, e.g. DuckDuckGo `https://duckduckgogg42xjoc72x3sjasowoarfbgcmvfimaftt6twagswzczad.onion`.
> 
> For now, we can only connect to ordinary websites. If we try to connect to an onion server, it will fail. In my case of using [LacticWhale/socks_dart](https://github.com/LacticWhale/socks_dart) to connect an onion router, it throws:
> ```
> SocketException: Failed host lookup: 'duckduckgogg42xjoc72x3sjasowoarfbgcmvfimaftt6twagswzczad.onion' (OS Error: No address associated with hostname, errno = 7)
> ```
> 
> By the way, does your [sneurlax/socks_socket](https://github.com/sneurlax/socks_socket) also provide a wayt to create or adapt a `HttpClient` or `Client`? Classes `HttpClient` and `Client` are from packages `dart:io` and `http` respectively. It'd be more user-friendly if it does. Some Flutter packages, e.g. Supabase, use `Client`.

I suspect, it's not using Tor Proxy for DNS resolution, which is why host lookup fails. It seems as though there is no solution that I am aware of

**Comment by nns52k** [comment]
@xAffan 
> I suspect, it's not using Tor Proxy for DNS resolution, which is why host lookup fails. It seems as though there is no solution that I am aware of

In my case, it's [LacticWhale/socks_dart](https://github.com/LacticWhale/socks_dart) that tries to do DNS lookup and fails, which behavior is well-described in [Tor Project: FAQ](https://2019.www.torproject.org/docs/faq.html.en#AccessOnionServices)
    Even though SOCKS 5 can accept either an IP or a hostname, most applications supporting SOCKS 5 *try to resolve the name before passing it to the SOCKS proxy.*

I've been looking for a Dart/Flutter package to support SOCKS 4a but it seems that there is none. SOCKS 4a only accepts hostname, so the package that support SOCKS 4a isn't likely to send DNS queries to Tor Proxy, because, again, SOCKS 4a only accept hostname.

**Comment by sneurlax** [comment]
@nns52k @xAffan did you resolve your issues with this?

**Comment by xAffan** [comment]
> @nns52k @xAffan did you resolve your issues with this?

You can connect with .onion sites just not with this plugin (presumably because it's disabled) other than that it is fine!

**Comment by sneurlax** [comment]
@xAffan I am now connecting to .onion nodes using* (basically) this plugin, but you have to edit `rust/Cargo.toml`'s `arti-client` entry with the `onion-service-client` option.  See https://github.com/cypherstack/tor/blob/main/rust/Cargo.toml#L17

**Comment by xAffan** [comment]
I'm not sure how to use plugins in flutter outside of pub add. Maybe, enable onion support but give a warning?

**Comment by sneurlax** [comment]
if it's still an issue I'll try to submit a PR to fix it for you @xAffan.  How's it working these days if you just pub add tor a fresh plugin?  or fork this and edit the example.  maybe we need to package in some dependency that's missing

**Comment by sneurlax** [comment]
Onion lookups should work with #52

**Comment by sneurlax** [comment]
Does this resolve your issue, @xAffan ?

**Comment by xAffan** [comment]
Yes my issue was resolved.
